enable recursive processing in kubectl edit

metral opened the pull request:
This PR was split out of https://github.com/kubernetes/kubernetes/pull/23673 per @deads2k's suggestion: https://github.com/kubernetes/kubernetes/pull/23673#discussion_r61291178

It makes use of the recursive processing of a directory in `kubectl edit`

metral commented:
@janetkuo can this PR get an "ok to test"?

janetkuo commented:
@k8s-bot ok to test

janetkuo commented:
@kubernetes/kubectl

metral commented:
@janetkuo anything else missing from this PR?

metral commented:
@janetkuo Do you happen to have an estimate as to when you can get to this PR? Thanks!

janetkuo commented:
@metral sorry for the late reply, reviewing this now.

janetkuo commented on pkg/kubectl/cmd/edit.go:
```diff
@@ -167,14 +169,22 @@ func RunEdit(f *cmdutil.Factory, out, errOut io.Writer, cmd *cobra.Command, args
-	clientConfig, err := f.ClientConfig()
+	infos := []*resource.Info{}
+	allErrs := []error{}
+	err = r.Visit(func(info *resource.Info, err error) error {
```
Why not use `r.Infos()` here?

metral replied:
Good catch - I was being creative on a previous attempt at getting things to work and I didn't realize I left the code to basically recreate `r.Infos()`, rather than enhancing it like I've done with other portions of the recursive feature.

I'll update it to use `r.Infos()`. Thanks

janetkuo commented on pkg/kubectl/cmd/edit.go:
```diff
@@ -184,7 +194,7 @@ func RunEdit(f *cmdutil.Factory, out, errOut io.Writer, cmd *cobra.Command, args
 	objs, err := resource.AsVersionedObjects(infos, defaultVersion.String(), encoder)
 	if err != nil {
-		return err
+		allErrs = append(allErrs, err)
```
Is it safe to continue even when `AsVersionedObjects` returns error?

metral replied:
I think you're right - I don't believe it is safe to continue. Can someone else verify this?

janetkuo commented on pkg/kubectl/cmd/edit.go:
```diff
@@ -211,11 +221,11 @@ outter:
 				w = crlf.NewCRLFWriter(w)
 			}
 			if err := results.header.writeTo(w); err != nil {
```
`writeTo` actually doesn't return error at all. Consider removing error from returned parameters.

metral replied:
will do

janetkuo commented on pkg/kubectl/cmd/edit.go:
```diff
@@ -211,11 +221,11 @@ outter:
 			if !containsError {
 				if err := printer.PrintObj(obj, w); err != nil {
-					return preservedFile(err, results.file, errOut)
+					allErrs = append(allErrs, preservedFile(err, results.file, errOut))
```
Same question here for continue with error

metral replied:
I believe this should be aggregated into `allErrs` because if a `PrintObj` fails during the editing of the file, then any remaining files in the directory will not be processed.

janetkuo commented:
cc @kargakis who's also familiar with `edit`

metral commented:
the e2e test seems to be having issues with one of its images not being able to be deleted on GCE: [log details](https://00e9e64bac7e114eef6f2fda2f35f4e557c0b4265caaf82b80-apidata.googleusercontent.com/download/storage/v1/b/kubernetes-jenkins/o/pr-logs%2Fpull%2F25085%2Fnode-pull-build-e2e-test%2F5578%2Fbuild-log.txt?qk=AD5uMEuUfAxG5GiCW82hE-QYMNk50vwD6TidADsY7dImvNuTGufXFlpfF0Uqvyvtgb5U_4rcqvf2BZzV0McL7xl9Z9grQhBKrhY7wchevFreL2UED3qJ2oveR_UhR9s8qN0ZDOgR3L1_NgvDPeN_mapzlT2pq0OEYwxRx-qHuQcMnRG6W-NwNKg-Se4mUVWIsYY4zLPflF857sAPc-DnwUh4ILWNM8ZryXc22GycTyGn6LqCs3BzHAY6vNQ5HxNpzknY1o9gLP3vb9ouVcOPx1EzeVKIaTR_9B2F1iBW1DTUT9v-nv3Mt9xXTRe9kZqHG_e3gGsCPRNt0W3bDnnDDa3IfxDq1rXvOO5Rt-QaQypfPgRKmj_ftviOSmdnrS9RIYfAqCyhrU5G43igV-WCREL_1ke1zKdB4Gvq4KYHU1a-k5reZAqLj-ArIYu0GVlmIQC96VgLHCZ1rf6paAZX6fim0Ll-jDRyhdiT7eG-TH_2GY-IzLpwgfl0v5Kc_XagJ4SzpE0498k1rtrFzQ7HJFSs77GDZwxXoK5NijX9iQpvF1_Jtfo8JGCXIXwBdLF4ntpDbQUaWxHQTq9lHwMv_0cmo1Wu_nWH716HpD59qvWFnT70Xj1qagTurTM3ClKjCKJ9sHWeJbiceCJWuWcBPb3yFStr70Fh_0Ls-bBKKxywJhKhw_IsbXr9U7hhP7Tw05yO2wLmXhGY_qzG0nGbgwb6T9qYM6-DUDWUggyG6XcjYvFzftm7wje17mGn_C6VNZhhTVtWC2GmDufUr-rtes2IfO28CfWu5WZX14RwIRKrynuxa7cNRACgLLf1zdYER_grLjCIyoe1ck9PBwxI1CO7JgSdUBTfjg)

metral commented:
@janetkuo I was trying to rerun the `GCE Node e2e` test due to the [error encountered](https://github.com/kubernetes/kubernetes/pull/25085#issuecomment-218802517) in GCE by closing/reopening the PR, and now the k8s-bot needs approval to retest even though it seems like the `GCE e2e` has restarted (kinda odd).

Can you please take a look when you get a chance to move this PR along? Thanks

kargakis commented:
@k8s-bot test this issue: https://github.com/kubernetes/kubernetes/issues/24479

kargakis commented:
@metral with this patch you effectively remove the ability from users to fix their errors when editinf a list. I would prefer if we fixed https://github.com/kubernetes/kubernetes/issues/20519 first (open PR: https://github.com/kubernetes/kubernetes/pull/20918) and then move forward with this.

metral commented:
@kargakis sgtm - I'll wait on #20918 to update this PR accordingly, thanks.

metral commented:
@k8s-bot test this issue: https://github.com/kubernetes/kubernetes/issues/25637

metral commented:
@kargakis I've rebased to include https://github.com/kubernetes/kubernetes/pull/20918 and squashed down.

>with this patch you effectively remove the ability from users to fix their errors when editing a list

Editing lists i.e. `kubectl edit namespaces` seems to be working on my end with this PR. Could you please specify what you mean?

Also, the Jenkin tests are failing due to: https://github.com/kubernetes/kubernetes/issues/25637

kargakis commented:
> Editing lists i.e. kubectl edit namespaces seems to be working on my end with this PR. Could you please specify what you mean?

What happens if you make an error when editing an object of the list?

metral commented:
@kargakis 

>What happens if you make an error when editing an object of the list?

I'm seeing the same output **with** & **without** this PR while editing lists. Here's the steps I'm doing for each:

I have two namespaces, `default` & `kube-system`:

**without** the PR I do the following:
- `kubectl edit namespaces` - this brings up editing for both namespaces in my editor (vim)
- the first namespace up for editing is `default`
- I intentionally throw in a typo into `apiVersion` (I spell it as just `Version`) & then I save/quit
- The editor gets reopened, now with the message `"# The edited file had a syntax error: unable to decode "edited-file": provided data does not appear to be a protobuf message, expected prefix [107 56 115 0]"`
- But I do nothing this time and I just save and quit again (with the typo still there), and it moves onto edit the next namespace `kube-system`
- I also do nothing with `kube-system` and just quit vim
- Edit returns: 
```
Edit cancelled, no valid changes were saved. // this is for default namespace
Edit cancelled, no changes made. // this is for kube-system namespace
```

**with** the PR, I do the same steps as above, and get the same exact output as above.

In both scenarios, if a user introduces an error upon the editing & saving of the object, it will reopen the editor with the failure message, as noted above, giving the user a second chance to correct their mistakes. if the user does not correct it this time & just saves and quits with the error still in place, it just gets restored to its original content and outputs the message: `Edit cancelled, no valid changes were saved`.

Am I missing something?

kargakis commented:
@metral oh you are appending errors that are returning anyway. Sorry, misunderstanding from my part. Anyway, the way `edit` manipulates lists is wrong and https://github.com/kubernetes/kubernetes/pull/20918 is about to merge so if you don't mind wait until that pull merges or cherry-pick it and work your changes on top. Thanks for being patient on this.

metral commented:
> oh you are appending errors that are returning anyway

@kargakis yes :)

> Anyway, the way edit manipulates lists is wrong and #20918 is about to merge so if you don't mind wait until that pull merges or cherry-pick it and work your changes on top. Thanks for being patient on this

I can certainly wait til https://github.com/kubernetes/kubernetes/pull/20918 gets merged - I'll keep an eye out for it to rebase & push this PR. Thanks for the help!

metral committed:
enable recursive processing in kubectl edit

metral commented:
@kargakis I cherry-picked https://github.com/kubernetes/kubernetes/pull/20918 and refactored my changes on top of it, but https://github.com/kubernetes/kubernetes/pull/20918 is stuck in the queue as its been having trouble, so it merging into master to then push my changes out has been a waiting game.

What do you recommend I do to ensure this gets into v1.3 as its the last piece of the original `--recursive` feature request per https://github.com/kubernetes/kubernetes/issues/19767?

deads2k commented:
@metral I haven't tagged #20918  as 1.3 because I could find a reference that someone thought it was required for 1.3.  The issue you linked  was a candidate, but not explicitly 1.3.  If you have something that indicates this is a 1.3 requirement I can mark a milestone on the other edit pull since it's a logical prereq.

metral commented:
@deads2k That makes sense. I don't have an indication that this PR is a requirement other than me wanting to get it merged into v1.3 for the sake of completeness wrt the rest of the `kubectl` subcommands that already support it because their PR's were merged.

I'll just follow your lead on https://github.com/kubernetes/kubernetes/pull/20918, as its a dep of this PR

metral commented:
@kargakis I've gone ahead and rebased on top of https://github.com/kubernetes/kubernetes/pull/20918

Just need an LGTM, thanks!

kargakis commented:
@k8s-bot test this issue: #IGNORE

metral commented:
@kargakis Is there anything else missing?

Also, `?w=1` is your friend here: https://github.com/kubernetes/kubernetes/pull/25085/files?w=1

kargakis commented:
LGTM, thanks!

kargakis commented:
@k8s-bot test this issue #IGNORE
